### PR TITLE
Split large batches on error instead of dropping them

### LINKED ISSUE #29778: Split large batches of documents if received 413 from Elasticsearch
**Describe the enhancement:**

Currently, after seeing a 413 response from Elasticsearch the whole batch is dropped and the error is logged (https://github.com/elastic/beats/pull/29368). Some of our customers would like to preserve at least some data from the batch instead of discarding the whole batch. See https://github.com/elastic/enhancements/issues/14589

The proposal is:

* When seeing a 413 response from Elasticsearch try to split the current batch (maybe in 2, or based on size of each document in the batch)
* If the 413 response is seen again – repeat the process until:
  * either all the smaller (split) batches are successfully sent to Elasticsearch or
  * the initial batch is reduced to a single document that cannot pass the `http.max_content_length` threshold in Elasticsearch
* If a batch contain only a single document that cannot be uploaded – drop the batch

Something similar was done in this PR https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/497

**Describe a specific use case for the enhancement or feature:**

Some of our clients are more sensitive to data loss than others and this enhancement would allow to preserve more data in case of misconfiguration of `http.max_content_length` in Elasticsearch or `bulk_max_size` in beats. **This would improve the situation in most of the cases but it would not completely solve the data loss problem**.

### LINKED ISSUE #34695: Shipper output fails on large events/batches
When event batches targeting the shipper exceed the RPC limit (currently 4MB), the shipper output drops all the events with an error similar to the following:

```json
{"log.level":"error","@timestamp":"2023-02-28T15:52:14.521Z","message":"failed to publish events: failed to publish the batch to the shipper, none of the 2048 events were accepted: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (4692625 vs. 4194304)","component":{"binary":"filebeat","dataset":"elastic_agent.filebeat","id":"filestream-monitoring","type":"filestream"},"log":{"source":"filestream-monitoring"},"log.origin":{"file.line":176,"file.name":"pipeline/client_worker.go"},"service.name":"filebeat","ecs.version":"1.6.0","log.logger":"publisher_pipeline_output","ecs.version":"1.6.0"}
```

There are multiple short-term mitigations (increase the RPC size limit; decrease the shipper output's batch size) but both of those approaches can still drop data unpredictably in the case of large events. If an event size is supported, then we should handle this error by splitting up the batch rather than permanently dropping its contents.

### PR DESCRIPTION
**Pull request by faec**
## What does this PR do?

Add a new method, `SplitRetry`, to `publisher.Batch`, which splits the events of the batch into two subsets and then separately retries each of them. Invoke this method in the Elasticsearch and Shipper outputs when we get an error indicating the batch was too large, instead of dropping them as previously.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### Elasticsearch output

- In `elasticsearch.yml`, set `http.max_content_length: 128kb` (smaller than that blocks all connections since it prevents filebeat from loading the initial index template)
- In `filebeat.yml`, set `output.elasticsearch.bulk_max_size: 500` and ingest some not-too-small log events.

When run at head, this should produce many errors of the form:

```json
{"log.level":"error","@timestamp":"2023-03-23T10:51:57.845-0400","log.logger":"elasticsearch","log.origin":{"file.name":"elasticsearch/client.go","file.line":241},"message":"failed to perform any bulk index operations: the bulk payload is too large for the server. Consider to adjust `http.max_content_length` parameter in Elasticsearch or `bulk_max_size` in the beat. The batch has been dropped","service.name":"filebeat","ecs.version":"1.6.0"}
```

and drops many events (in my test it only ingested 1536 out of 1 million in my test logs).

When run with this PR, these errors should be gone, and it should ingest 100% of the events.

### Shipper output

- Rebuild the shipper with an artificially low RPC limit by editing [inputhandler.go](https://github.com/elastic/elastic-agent-shipper/blob/a526fd696ec6444600abb718caa2dd04798beb8a/grpcserver/inputhandler.go#L53) to use e.g. `64 * units.KiB` rather than `64 * units.MiB`
- Try ingesting into the Shipper (default Agent config should be enough when pointed out the shipper, the standard metrics should be big enough to trigger the issue)

When run at head, some events should be dropped, and the logs should show many errors similar to:

```json
"dropping 50 events because of RPC failure: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (118688 vs. 65536)"
```

When run with this PR, the errors should be gone and ingestion should succeed (as long as no individual event exceeds the RPC limit). For me the typical batch size in the logs drops from 50 to 25 or lower.

## Related issues

- Closes #29778 
- Closes #34695

### CODE REVIEW

**Review by leehinman — CHANGES_REQUESTED** [review]
Looks good.  Can we add 2 more test cases.  One when you split but all messages are too big so all the retries fail.  And the second when you split and one batch works but the other fails because it is too big.

**Comment by faec** [comment]
> Looks good. Can we add 2 more test cases. One when you split but all messages are too big so all the retries fail. And the second when you split and one batch works but the other fails because it is too big.

Hmm, some obstacles to meaningfully testing these scenarios:
- Nothing in these tests exercises actual size, the mocked RPC just returns simulated success or failure based on test case config. We could expand the tests so the mocked server returns different statuses on subsequent calls, but:
- `Publish` itself is stateless and has no way of telling whether the batch it's given is split, or being retried, or otherwise has any sequential significance. The only mechanism of carryover is the retryer, which in a real pipeline would be listening on the retry channel the tests use to verify `Publish` behavior. But as far as `Publish` itself is concerned the only thing we can meaningfully check is whether it handles an individual batch correctly based on the resulting RPC status code, and the tests already check every currently handled combination of status code / `SplitRetry` success.

To the extent that nested success / failure can be tested, it's currently done in the `SplitRetry` tests, which check recursive splitting and final batch acknowledgment.

File: libbeat/outputs/elasticsearch/client.go [hunk 1]
```diff
@@ -189,8 +189,25 @@ func (client *Client) Publish(ctx context.Context, batch publisher.Batch) error
...
 
 	switch {
-	case err == errPayloadTooLarge:
-		batch.Drop()
+	case errors.Is(err, errPayloadTooLarge):
+		if batch.SplitRetry() {
+			// If we successfully split the batch, we report the events to the
+			// observer as "failed", since this indicates a publish error that
+			// will be retried.
+			client.observer.Failed(len(events))
```
Comment (cmacknz) [inline libbeat/outputs/elasticsearch/client.go:197]: Is it worth adding a metric like `splits` for the number of split batches? Do we have any way to know if batch splitting is happening or possibly causing problems?

Probably to debug any issues from batch splitting we'd need to look at the events and batches directly, but having a counter or some way to know if it has ever happened would give us a quick way to rule it out.

Comment (faec) [reply libbeat/outputs/elasticsearch/client.go]: Done, added new metric `batches.split` alongside `events.{failed, dropped, etc}`.

**Comment by faec** [comment]
> I can be convinced that this isn't the right place to do the test. But I still think we need to test what happens in these edge cases. Is there an existing publish test that we could expand?

The main difficulty is that meaningfully testing these scenarios requires a full running pipeline (with retryer, output workers, and the rest) -- since `Publish` itself is ~stateless what is really being tested here is full pipeline retry operation, which is closer to an end-to-end test. This gets pretty tricky -- we would need enough of the pipeline running that it has full autonomous retry behavior, but not so much that it creates real connections or introduces timing-based nondeterminism into batch/reply order.

I can spend some time looking for a better compromise, but could you say a little more about what specific logic or data flow you want tested? Would you be satisfied by a slightly augmented unit test where the test code itself manually sends the "retried" values through `Publish` again, or is the full-pipeline retry behavior part of what you're getting at?

File: CHANGELOG.next.asciidoc [hunk 2]
```diff
@@ -59,6 +59,8 @@ https://github.com/elastic/beats/compare/v8.2.0\...main[Check the HEAD diff]
 - Fix race condition when stopping runners {pull}32433[32433]
 - Fix concurrent map writes when system/process code called from reporter code {pull}32491[32491]
 - Fix panics when a processor is closed twice {pull}34647[34647]
+- Update elastic-agent-system-metrics to v0.4.6 to allow builds on mips platforms. {pull}[]
```
Comment (cmacknz) [inline CHANGELOG.next.asciidoc:62]: Looks like an extra changelog entry was introduced.

**Comment by cmacknz** [comment]
This looks good to me, pending resolution of of Lee's requests for additional tests.

I like how small and easy to follow this implementation turned out to be, although I'm sure a lot of thought went into make it this way. Thanks!

**Review by leehinman — APPROVED** [review]
awesome.  thank you.

**Comment by faec** [comment]
Looks like the failures are unrelated tests that couldn't download from github.com and nodejs.org -- presumably flakiness in the test environment.

**Pull request merged**